File: python/pokerTest.py
import sys
import os
from random import shuffle
from Cards import Cards
from Cards import Deck
from Cards import Hand
from Cards import Player
from Cards import PlayerRank

def compare_suits(suit1, suit2):
    suit_order = "DCHS"  # The order is Diamonds, Clubs, Hearts, Spades
    if suit1 is None:
        return 1  # None is considered greater than any suit
    if suit2 is None:
        return -1  # None is considered greater than any suit
    return suit_order.index(suit1) - suit_order.index(suit2)
    
# Function to read a deck from a file
def read_deck_from_file(file_path):
    all_hands = []
    try:
        with open(file_path, 'r') as reader:
            lines = reader.readlines()
            unique_cards = set()
            print("File Name: " + file_path)
            for line in lines:
                print(line.strip())
                card_records = line.strip().split(',')
                if len(card_records) != 5:
                    print("Invalid card record: " + str(card_records))
                    print(len(card_records))
                    continue
                for card_record in card_records:
                    if card_record in unique_cards:
                        print(" *** ERROR - DUPLICATE CARD FOUND IN DECK *** ")
                        print("DUPLICATE: " + card_record)
                        sys.exit(1)
                    unique_cards.add(card_record)
                if len(card_records) == 5:
                    new_hand = []
                    for card_str in card_records:
                        if len(card_str) == 2:
                            rank = card_str[0]
                            suit = card_str[1]
                        elif len(card_str) == 3:
                            rank = card_str[:2]
                            suit = card_str[2]
                        else:
                            print("Invalid card format: " + card_str)
                            continue
                        card = Cards(rank, suit)
                        new_hand.append(card)
                    all_hands.append(new_hand)
    except IOError as e:
        print("Error reading the file: " + str(e))
        sys.exit(1)

    if len(all_hands) < 5:
        print("Not enough cards to make a hand.")
    else:
        print("Here are the six hands...")
        for i in range(6):
            hand = all_hands[i]
            print("Hand " + str(i + 1) + ":")
            for card in hand:
                print(str(card), "\t")
            print()

if __name__ == "__main__":
    deck = Deck()
    deck.shuffle()

    if len(sys.argv) == 1:
        print("*** USING RANDOMIZED DECK OF CARDS ***")
        print()
        print()
        print("*** Shuffled 52 card deck: ")
    else:
        print("*** USING TEST DECK ***")  
        file = sys.argv[1]
        read_deck_from_file(file)
        sys.exit(0)

    num_players = 6
    cards_per_hand = 5
    new_hand = Hand()
    deck.print_deck()

    # Creating and dealing hands to players
    players = [Player("Player " + str(i + 1)) for i in range(num_players)]

    # Making a list to store objects of the class PlayerRank
    player_ranks = []

    for i in range(cards_per_hand):
        for player in players:
            new_card = deck.deal_card()
            if new_card is not None:
                player.add_to_hand(new_card)
            else:
                print("The deck is empty.")
                break

    player_ranks.sort()

    # Calculating and storing ranks for each player's hand
    print("Here are the six hands... ")
    for new_player in players:
        print(new_player.get_name() + " 's Hand:")
        new_player.get_hand().print_hand()
        hand_ranks = new_player.get_hand().rank_hand()
        flush_suit = new_player.get_hand().get_flush_suit()
        print()
        player_ranks.append(PlayerRank(new_player, hand_ranks, flush_suit))

    # Sorting players in order of first, then second, then third, etc...
    player_ranks.sort(key=lambda pr: (-Cards.get_rank_value(pr.get_rank()), compare_suits(pr.get_suit(), pr.get_suit())))

    # Printing out each hand and rank in order
    print("//////////////////////////Winning Hand Order//////////////////////////////////////////////////////")
    for new_player_rank in player_ranks:
        print(new_player_rank.get_player().get_name() + " 's Hand:")
        new_player_rank.get_player().get_hand().sort_hand()
        print(" - " + new_player_rank.get_rank())
        print()

    deck.print_remaining_deck()

File: rust/pokerTest.rs
#!/usr/bin/rust
use std::collections::{HashSet, HashMap};
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let mut deck = Deck::new();
    deck.shuffle();

    let args: Vec<String> = std::env::args().collect();

    if args.len() == 1 {
        println!("*** P O K E R   H A N D   A N A L Y Z E R ***");
        println!();
        println!("*** USING RANDOMIZED DECK OF CARDS ***");
        println!();
        println!("*** Shuffled 52 card deck: ");
    } else {
        println!("*** P O K E R   H A N D   A N A L Y Z E R ***");
        println!();
        println!("*** USING TEST DECK ***");
        let file = &args[1];
        read_deck_from_file(file);
        return;
    }

    let num_players = 6;
    let cards_per_hand = 5;
    let mut players: Vec<Player> = (1..=num_players).map(|i| Player::new(format!("Player {}", i))).collect();
    let mut player_ranks: Vec<PlayerRank> = Vec::new();

    for _ in 0..cards_per_hand {
        for player in players.iter_mut() {
            if let Some(new_card) = deck.dealing_cards() {
                player.adding_to_hand(new_card);
            } else {
                println!("The deck is empty.");
                break;
            }
        }
    }

    player_ranks.sort();

    println!("Here are the six hands... ");
    for new_player in players.iter() {
        new_player.get_hand().print_hand();
        let hand_ranks = new_player.get_hand().rank_hand();
        let flush_suit = new_player.get_hand().get_flush_suit();
        println!();
        player_ranks.push(PlayerRank::new(new_player.clone(), hand_ranks, flush_suit));
    }

    player_ranks.sort_by(|pr1, pr2| {
        let rank_compare = pr2.get_rank().cmp(pr1.get_rank());

        if rank_compare == std::cmp::Ordering::Equal {
            if pr1.get_rank() == "Straight" && pr2.get_rank() == "Straight" {
                return compare_suits(pr1.get_suit(), pr2.get_suit());
            } else if pr1.get_rank() == "Two Pair" && pr2.get_rank() == "Two Pair" {
                let kicker_compare = compare_suits(pr1.get_suit(), pr2.get_suit());
                if kicker_compare == std::cmp::Ordering::Equal {
                    return pr1.get_player().get_name().cmp(pr2.get_player().get_name());
                }
                return kicker_compare;
            } else if pr1.get_rank() == "Pair" && pr2.get_rank() == "Pair" {
                return compare_suits(pr1.get_suit(), pr2.get_suit());
            } else if pr1.get_rank() == "High Card" && pr2.get_rank() == "High Card" {
                let high_card = compare_suits(pr1.get_suit(), pr2.get_suit());
                if high_card == std::cmp::Ordering::Equal {
                    return pr1.get_player().get_name().cmp(pr2.get_player().get_name());
                }
            } else {
                let suit_compare = compare_suits(pr1.get_suit(), pr2.get_suit());
                if suit_compare == std::cmp::Ordering::Equal {
                    return pr1.get_player().get_name().cmp(pr2.get_player().get_name());
                }
                return suit_compare;
            }
        }

        return rank_compare;
    });

    deck.print_remaining_deck();
    println!();
    println!();
    println!("--- WINNING HAND ORDER ---");
    for new_player_rank in player_ranks.iter() {
        new_player_rank.get_player().get_hand().sort_hand();
        new_player_rank.get_player().get_hand().print_hand();
        println!(" - {}", new_player_rank.get_rank());
        println!();
    }
}

// Comparing suits in case of a tie
fn compare_suits(suit1: char, suit2: char) -> std::cmp::Ordering {
    let suit_order = "DCHS"; // The order is Diamonds, Clubs, Hearts, Spades
    return suit_order.chars().position(|c| c == suit1).unwrap().cmp(&suit_order.chars().position(|c| c == suit2).unwrap());
}

// Method for reading in the file to create the hands and print them
fn read_deck_from_file(file_path: &str) {
    let path = Path::new(file_path);
    let file = match File::open(&path) {
        Err(_) => {
            println!("Error reading the file.");
            return;
        }
        Ok(file) => file,
    };

    let reader = io::BufReader::new(file);
    let mut all_hands: Vec<Vec<Cards>> = Vec::new();
    let mut unique_cards: HashSet<String> = HashSet::new();

    println!("File Name: {}", file_path);
    for line in reader.lines() {
        let line = match line {
            Err(_) => continue,
            Ok(line) => line,
        };

        println!("{}", line);

        let card_records: Vec<&str> = line.split(",").collect();
        if card_records.len() != 5 {
            println!("Invalid card record: {}", card_records.join(","));
            continue;
        }

        for card_record in &card_records {
            if unique_cards.contains(*card_record) {
                println!("*** ERROR - DUPLICATE CARD FOUND IN DECK ***");
                println!("DUPLICATE: {}", card_record);
                std::process::exit(1);
            }
            unique_cards.insert(card_record.to_string());
        }

        if card_records.len() == 5 {
            let new_hand: Vec<Cards> = card_records
                .iter()
                .map(|&card_str| {
                    let (rank, suit) = if card_str.len() == 2 {
                        (card_str.get(0..1).unwrap(), card_str.get(1..2).unwrap())
                    } else if card_str.len() == 3 {
                        (card_str.get(0..2).unwrap(), card_str.get(2..3).unwrap())
                    } else {
                        println!("Invalid card format: {}", card_str);
                        return Cards::new("", ' ');
                    };
                    Cards::new(rank, suit.chars().next().unwrap())
                })
                .collect();
            all_hands.push(new_hand);
        }
    }

    if all_hands.len() < 5 {
        println!("Not enough cards to make a hand.");
    } else {
        println!();
        println!("*** Here are the six hands...");
        for i in 0..6 {
            let hand = &all_hands[i];
            for card in hand {
                print!("{}\t", card);
            }
            println!();
        }
    }
}